### PR TITLE
adjust guide-kafka

### PR DESCRIPTION
**Pull request by lilliCao**
Implements:
 * move content of `guide-kafka` to `/spring/guide-kafka-spring`
 * add link to quarkus kafka guide and to `/spring/guide-kafka-spring`

Closes https://github.com/devonfw/devon4quarkus/issues/8

### CODE REVIEW

File: documentation/guide-kafka.asciidoc [hunk 1]
```diff
@@ -6,393 +6,7 @@ toc::[]
...
-[source,java]
-----
- @Bean
-  @ConfigurationProperties(prefix = "messaging.kafka.producer")
-  public KafkaProducerProperties messageKafkaProducerProperties() {
-    return new KafkaProducerProperties();
-  }
-----
-
-For producer and consumer the prefixes are `messaging.kafka.producer...` and `message.kafka.consumer...`` and for retry the prefix is `messaging.retry...``
```
Comment (GuentherJulian) [inline documentation/guide-kafka.asciidoc:73]: One backtick too much after `message.kafka.consumer...` and `messaging.retry...`

File: documentation/guide-kafka.asciidoc [hunk 2]
```diff
@@ -6,393 +6,7 @@ toc::[]
...
-* The `https://github.com/devonfw/devon4j/blob/develop/modules/kafka/src/main/java/com/devonfw/module/kafka/common/messaging/api/config/KafkaConsumerProperties.java/[Devon4j-Kafka-Consumer properties]` are defined from `https://kafka.apache.org/documentation/#consumerconfigs/[kafka consumer config]`.
-* The `https://github.com/devonfw/devon4j/blob/develop/modules/kafka/src/main/java/com/devonfw/module/kafka/common/messaging/api/config/KafkaListenerContainerProperties.java/[Devon4j-Kafka-Listener container properties]` are defined from `https://docs.spring.io/spring-kafka/api/org/springframework/kafka/listener/ContainerProperties.html/[Spring kafka listener properties]`.
-
-
-== Naming convention for topics
-
-For better managing of several Kafka topics in your application portfolio we devon strongly advice to introduce a naming scheme for your topics. The schema may depend on the actual usage pattern of kafka. For context where kafka is used
```
Comment (GuentherJulian) [inline documentation/guide-kafka.asciidoc:86]: Remove the "devon" in "... we devon strongly advice ..."

File: documentation/guide-kafka.asciidoc [hunk 3]
```diff
@@ -6,393 +6,7 @@ toc::[]
...
-  }
-----
-
-The group id can be mentioned in `application.properties` as listener properties.
-
-[source,properties]
-----
-messaging.kafka.consumer.groupId=default
-----
-if there are multiple topics and multiple listeners then we suggest to specify the topic names directly on each listeners instead reading from the property file.
```
Comment (GuentherJulian) [inline documentation/guide-kafka.asciidoc:164]: If with initial capital

File: documentation/guide-kafka.asciidoc [hunk 4]
```diff
@@ -6,393 +6,7 @@ toc::[]
 Messaging Services provide an asynchronous communication mechanism between applications. Technically this is
 implemented using http://kafka.apache.org/documentation.html/[Apache Kafka] .
 
-Devonfw uses link:https://spring.io/projects/spring-kafka[Spring-Kafka] as kafka framework.
+For link:spring.asciidoc[spring], devonfw uses link:https://spring.io/projects/spring-kafka[Spring-Kafka] as kafka framework.
+For more details, check the link:spring/guide-kafka-spring.asciidoc[devon4j-kafka].
 
-This guide explains how Spring kafka is used in devonfw applications. It focuses on aspects which are special to devonfw if you want to learn about spring-kafka you should adhere to springs references documentation.
```
Comment (GuentherJulian) [inline documentation/guide-kafka.asciidoc:11]: Maybe we can add some general words about Kafka in this guide. Also the parts about naming conventions or parallel processing are not really spring or devon4j specific.

**Review by GuentherJulian — CHANGES_REQUESTED** [review]

File: documentation/guide-kafka.asciidoc (hunk 4 above)
Comment (lilliCao) [reply documentation/guide-kafka.asciidoc:11]: I see but I find it difficult for user to follow if we just seperate those aspects.

**Review by GuentherJulian — APPROVED** [review]

File: documentation/spring/guide-kafka-spring.asciidoc [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+messaging.retry.back-off-policy.retryMaxDelay.default=600000
+messaging.retry.back-off-policy.retryCount.default=2
+
+# Default retry policy properties
+messaging.retry.retry-policy.retryPeriod.default=1800
+messaging.retry.retry-policy.retryableExceptions.default=<Class names of exceptions for which a retry should be performed>
+messaging.retry.retry-policy.retryableExceptionsTraverseCauses.default=true
+-----
+
+By giving properties like above , the same values will be passed for all the topics and the way of processing retry for all the topics are same.
```
Comment (EduardKrieger) [inline documentation/spring/guide-kafka-spring.asciidoc:275]: Remove the space in front of the comma

File: documentation/spring/guide-kafka-spring.asciidoc [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+
+All these above property values are mapped to the classes `DefaultBackOffPolicyProperties.java` and `DefaultRetryPolicyProperties.java` and configured by the class `MessageDefaultRetryConfig.java`.
+
+The MessageRetryContext in devon kafka is used to perform the retry pattern with the properties from DefaultBackOffPolicyProperties and DefaultRetryPolicyProperties.
+
+The 2 main properties of MessageRetryContext is nextRetry and retryUntil which is a `Instant` date format and it is calculated internally using the properties given in DefaultBackOffPolicyProperties and DefaultRetryPolicyProperties.
```
Comment (EduardKrieger) [inline documentation/spring/guide-kafka-spring.asciidoc:281]: are nextRetry and retryUntil ...

File: documentation/spring/guide-kafka-spring.asciidoc [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+
+You may change the behavior of this date calculation by providing your own implementation classes for `MessageBackOffPolicy.java` and `MessageRetryPolicy.java`.
+
+The naming convention for retry topic is the same topic name which you have given to publish the message and we add suffix `-retry` to it once it is consumed and given to process with retry.
+
+If there is no topic found in the consumed record the default retry topic will be added which is `default-message-retry`.
+
+=== Retry topics
+Devon4j-kafka uses a separate retry topic for each topic where retries occur. By default this topic is named `<topic name>-retry`. You may change this behavior by providing your own implementation for `DefaultKafkaRecordSupport` which is an default implementation from devon4j-kafka for `KafkaRecordSupport`.
```
Comment (EduardKrieger) [inline documentation/spring/guide-kafka-spring.asciidoc:290]: which is a default ...

File: documentation/spring/guide-kafka-spring.asciidoc [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+=== Handling retry finally failed
+
+Per default when the retry fails with final attempt  we just log the message and delete the payload of ProducerRecord which comes to proceed the retry pattern.
+
+You can change this behavior by providing the implementation class for the interface `MessageRetryHandler.java`
+which has two method `retryTimeout` and `retryFailedFinal`.
+
+== Tracer
+We leverage https://spring.io/projects/spring-cloud-sleuth[Spring Cloud Sleuth] for tracing in devon4j-kafka
+This is used to trace the asynchronous process of kafka producing and consuming. In an asynchronous process it is important to maintain a id which will be same for all asynchronous process.
```
[the remark this answers is not in the document]
Comment (EduardKrieger) [reply documentation/spring/guide-kafka-spring.asciidoc:303]: process,

[the remark this answers is not in the document]
Comment (EduardKrieger) [reply documentation/spring/guide-kafka-spring.asciidoc:303]: it is important to maintain an id that will be the same for all asynchronous processes

File: documentation/spring/guide-kafka-spring.asciidoc [hunk 9]
```diff
@@ -0,0 +1,393 @@
...
+
+This is a part of both starter dependencies `starter-kafka-receiver` and `starter-kafka-sender`.
+
+There are 2 important properties which will be automatically logged which is trace-id and spain-id.
+The trace-id is same for all the asynchronous process and span-id is unique for each asynchronous process.
+
+=== How devon4j-kafka handles tracer ?
+
+We inject the trace-id and span-id in to the ProducerRecord headers which comes to publish into the kafka broker.
+Its injected in the headers with the key `traceId` for trace-id and `spanId` for span-id.
```
[the remark this answers is not in the document]
Comment (EduardKrieger) [reply documentation/spring/guide-kafka-spring.asciidoc:314]: It is

File: documentation/spring/guide-kafka-spring.asciidoc [hunk 10]
```diff
@@ -0,0 +1,393 @@
...
+== Logging
+devon4j-kafka provides multiples support classes to log the published message and the consumed message.
+* The class `ProducerLoggingListener` which implements ProducerListener<K,V> from spring kafka uses to log the message as soon as it is published in the kafka broker.
+
+* The aspect class `MessageListenerLoggingAspect` which is annotated with `@Aspect` and has a method `logMessageprocessing` which is annotated with `@Around("@annotation(org.springframework.kafka.annotation.KafkaListener)&&args(kafkaRecord,..)")`
+used to listen to the classes which is annotated with `@KafkaListener` and logs the message as soon as it is consumed.
+
+* The class `MessageLoggingSupport` has multiple methods to log different type of events like MessageReceived, MessageSent, MessageProcessed, MessageNotProcessed.
+
+* The class `LoggingErrorHandler` which implements `ErrorHandler` from spring-kafka which logs the message when an error occurred while consuming message. You may change this behavior by creating your own implementation class for the ErrorHandler.
```
[the remark this answers is not in the document]
Comment (EduardKrieger) [reply documentation/spring/guide-kafka-spring.asciidoc:330]: while consuming messages (or a message)

File: documentation/spring/guide-kafka-spring.asciidoc [hunk 11]
```diff
@@ -0,0 +1,393 @@
...
+
+== Kafka Health check using Spring acutator
+The spring config class MessageCommonConfig automatically provides a spring health indicator bean for kafka if
+the property endpoints. The health indicator will check for all topics listed in `messaging.kafka.health.topics-tocheck`
+if a leader is available. If this property is missing only the broker connection will be checked. The timeout for
+the check (default 60s) maybe changed via the property `messaging.kafka.health.timeout`.
```
[the remark this answers is not in the document]
Comment (EduardKrieger) [reply documentation/spring/guide-kafka-spring.asciidoc:336]: Kafka

File: documentation/spring/guide-kafka-spring.asciidoc [hunk 12]
```diff
@@ -0,0 +1,393 @@
...
+  public void consumer(ConsumerRecord<K, V> consumerRecord, Acknowledgment acknowledgment) {
+...
+    }
+  }
+-----
+
+With this annotation in-place each message will be checked for a valid JWT in a message header with the name `Authorization`. If a valid annotation is found the spring security context will be initialized with the user roles and "normal" authorization e.g. with `@RolesAllowed` may be used. This is also demonstrated in the kafka sample application.
+
+== Using Kafka for internal parallel processing
+Apart from the use of Kafka as "communication channel" it sometimes helpful to use Kafka internally to do parallel processing:
```
Comment (EduardKrieger) [inline documentation/spring/guide-kafka-spring.asciidoc:383]: it is sometimes helpful

File: documentation/spring/guide-kafka-spring.asciidoc [hunk 13]
```diff
@@ -0,0 +1,393 @@
...
+
+.Architecture for internal parallel processing with Kafka
+image::../images/kafka-architecture-internal.png["Architecture for internal parallel processing with Kafka",scaledwidth="80%",align="center",link="https://devonfw.com/website/pages/docs/images/kafka-architecture-internal.svg"]
+
+This examples shows a payment service which allows a to submit a list of receipt IDs for payment.
+We assume that the payment it self takes a long time and should be done asynchronously and in parallel.
```
Comment (EduardKrieger) [inline documentation/spring/guide-kafka-spring.asciidoc:389]: itself

**Review by EduardKrieger — CHANGES_REQUESTED** [review]

File: documentation/guide-kafka.asciidoc [hunk 14]
```diff
@@ -6,393 +6,7 @@ toc::[]
 Messaging Services provide an asynchronous communication mechanism between applications. Technically this is
 implemented using http://kafka.apache.org/documentation.html/[Apache Kafka] .
 
-Devonfw uses link:https://spring.io/projects/spring-kafka[Spring-Kafka] as kafka framework.
+For link:spring.asciidoc[spring], devonfw uses link:https://spring.io/projects/spring-kafka[Spring-Kafka] as kafka framework.
+For more details, check the link:spring/guide-kafka-spring.asciidoc[devon4j-kafka].
```
[the remark this answers is not in the document]
Comment (TugbaDalmaz) [reply documentation/guide-kafka.asciidoc:9]: Kafka

File: documentation/guide-kafka.asciidoc (hunk 2 above)
Comment (TugbaDalmaz) [reply documentation/guide-kafka.asciidoc:86]: Kafka

File: documentation/guide-kafka.asciidoc (hunk 4 above)
Comment (TugbaDalmaz) [reply documentation/guide-kafka.asciidoc:11]: Kafka

**Review by EduardKrieger — APPROVED** [review]

**Review by EduardKrieger — APPROVED** [review]

**Pull request merged**